Batteries/Data/Fin/Lemmas.lean

```diff
@@ -42,63 +42,70 @@ theorem findSome?_succ_of_none {f : Fin (n+1) → Option α} (h : f 0 = none) :
+      exists_and_left]
 
-theorem eq_none_of_findSome?_eq_none {f : Fin n → Option α} (h : findSome? f = none) (i) :
-    f i = none := by
-  induction n with
-  | zero => cases i; contradiction
-  | succ n ih =>
-    rw [findSome?_succ] at h
-    match heq : f 0 with
-    | some x =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_some] at h
-      contradiction
-    | none =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_none] at h
-      cases i using Fin.cases with
-      | zero => exact heq
-      | succ i => exact ih h i
-
-@[simp] theorem findSome?_isSome_iff {f : Fin n → Option α} :
-    (findSome? f).isSome ↔ ∃ i, (f i).isSome := by
-  simp only [Option.isSome_iff_exists]
-  constructor
-  · intro ⟨x, hx⟩
-    match exists_of_findSome?_eq_some hx with
-    | ⟨i, hi⟩ => exists i, x
-  · intro ⟨i, x, hix⟩
-    match h : findSome? f with
-    | some x => exists x
-    | none => rw [eq_none_of_findSome?_eq_none h i] at hix; contradiction
+@[simp]
+theorem findSome?_isSome_iff {f : Fin n → Option α} :
+    (findSome? f).isSome ↔ ∃ i, (f i).isSome ∧ ∀ j < i, (f j).isNone := by
+  simp only [Option.isSome_iff_exists, findSome?_eq_some_iff, exists_comm, exists_and_right,
+    Option.isNone_iff_eq_none]
 
 @[simp] theorem findSome?_eq_none_iff {f : Fin n → Option α} :
     findSome? f = none ↔ ∀ i, f i = none := by
-  constructor
-  · exact eq_none_of_findSome?_eq_none
-  · intro hf
-    match h : findSome? f with
-    | none => rfl
-    | some x =>
-      match exists_of_findSome?_eq_some h with
-      | ⟨i, h⟩ => rw [hf] at h; contradiction
+  induction n with
+  | zero =>
+    simp only [findSome?_zero, true_iff]
+    exact fun i => i.elim0
+  | succ n ih =>
+    simp only [findSome?_succ, Option.orElse_eq_orElse,
+      Option.orElse_eq_or, Option.or_eq_none_iff, ih, forall_fin_succ]
 
 theorem findSome?_isNone_iff {f : Fin n → Option α} :
     (findSome? f).isNone ↔ ∀ i, (f i).isNone := by simp
 
+theorem exists_of_findSome?_eq_some {f : Fin n → Option α} (h : findSome? f = some x) :
+    ∃ i, f i = some x ∧ ∀ j < i, f j = none := by
+  rwa [findSome?_eq_some_iff] at h
+
+theorem eq_none_of_findSome?_eq_none {f : Fin n → Option α} (h : findSome? f = none) (i) :
+    f i = none := by
+  rw [findSome?_eq_none_iff] at h
+  apply h
+
+theorem exists_of_findSome?_isSome {f : Fin n → Option α} (h : (findSome? f).isSome) :
+    ∃ i, (f i).isSome ∧ ∀ j < i, (f j).isNone := by
+  rwa [findSome?_isSome_iff] at h
+
+theorem isNone_of_findSome?_isNone {f : Fin n → Option α} (h : (findSome? f).isNone) :
+  (f i).isNone := by
+  simp only [Option.isNone_iff_eq_none, findSome?_eq_none_iff] at h ⊢
+  exact h _
+
+theorem isSome_findSome?_of_isSome {f : Fin n → Option α} (h : (f i).isSome) :
+    (findSome? f).isSome := by
+  simp only [Option.isSome_iff_ne_none, ne_eq, findSome?_eq_none_iff]
+  grind
+
+
+@[deprecated findSome?_isSome_iff (since := "2025-09-26")]
```

Comment:
It's not useful to deprecate a theorem that never existed.

Reply:
Good point. I de-deprecated it anyway, I can see reasons to have both.